driver/reducecalculation_driver.hpp

```diff
@@ -237,27 +312,51 @@ int ReduceCalculationDriver<Tgpu, Tref>::AllocateBuffersAndCopy()
     uint32_t ctx = 0;
 
     in_dev        = std::unique_ptr<GPUMem>(new GPUMem(ctx, in_sz, sizeof(Tgpu)));
-    out_dev       = std::unique_ptr<GPUMem>(new GPUMem(ctx, out_sz, sizeof(Tgpu)));
+    out_dev       = IsLogicalCalculation() ? std::make_unique<GPUMem>(ctx, out_sz, sizeof(bool))
+                                           : std::make_unique<GPUMem>(ctx, out_sz, sizeof(Tgpu));
     workspace_dev = std::unique_ptr<GPUMem>(new GPUMem(ctx, ws_sizeInBytes, sizeof(std::byte)));
 
-    in      = std::vector<Tgpu>(in_sz, static_cast<Tgpu>(0));
-    out     = std::vector<Tgpu>(out_sz, static_cast<Tgpu>(0));
-    outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+    in = std::vector<Tgpu>(in_sz, static_cast<Tgpu>(0));
 
-    for(int i = 0; i < in_sz; ++i)
+    if(IsLogicalCalculation())
     {
-        in[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(0.0), static_cast<Tgpu>(1.0));
+        logical_out     = std::vector<uint8_t>(out_sz, 0);
+        logical_outhost = std::vector<uint8_t>(out_sz, 0);
+    }
+    else
+    {
+        out     = std::vector<Tgpu>(out_sz, static_cast<Tgpu>(0));
+        outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+    }
+
+    for(size_t i = 0; i < in_sz; ++i)
+    {
+        bool is_zero = prng::gen_A_to_B<int>(0, 2) == 0;
+        in[i] =
+            is_zero ? 0 : prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(0.0), static_cast<Tgpu>(2.0));
     }
 
     if(in_dev->ToGPU(GetStream(), in.data()) != 0)
     {
         std::cerr << "Error copying (in) to GPU, size: " << in_dev->GetSize() << std::endl;
         return miopenStatusAllocFailed;
     }
-    if(out_dev->ToGPU(GetStream(), out.data()) != 0)
+
+    if(IsLogicalCalculation())
     {
-        std::cerr << "Error copying (out) to GPU, size: " << out_dev->GetSize() << std::endl;
-        return miopenStatusAllocFailed;
+        if(out_dev->ToGPU(GetStream(), logical_out.data()) != 0)
+        {
+            std::cerr << "Error copying (out) to GPU, size: " << out_dev->GetSize() << std::endl;
+            return miopenStatusAllocFailed;
```

Comment:
should change return status code

Reply:
solved

Reply:
should we also change return status code for `if(ws_sizeInBytes == static_cast<size_t>(-1))`?

Reply:
I think status code for this is fine